### DISCUSSION TITLE
Gantt charts for state machines- An awsome overall view of a process.

### DISCUSSION THREAD

**Discussion by mariorl** [frangoteam/FUXA#848] [Ideas]
Hi,
Usually when you start a complex automation task, state machines are very useful. In this way, every state of the process is a value of an integer variable. So the process starts and continues through different stages, and every stage  is a new value. Sometimes go back so that the process var takes a previous value again. When you assign every possible value to a one single color, and put the time on the x axis, you get an awsome view of what is happening in a quick view. You will see if the process run forward normally, if it went back in a certain moment, if it has been aborted, if the time of every stage is the usual... 
Other very useful feature to use combined with the previous is the ability of represent digital traces. Today I do this in Scadas using an analog trace in a trend graph, asigning "false" to a value in a way the analog trace is placed at a certain height in the trend chart, between other digital traces, and "true" a little higher, still between the previous and de next digital traces. With this you create an horizontal zone at a certain height who emulates a digital channel. there are very few, if any, commercial solutions that can do this.

**Comment by unocelli** [comment]
Hi, 
I am not sure I understand your request. the control used in FUXA for the chart is [uplot](https://github.com/leeoniya/uPlot), you can refer to one of the examples.

**Reply by mariorl** [reply]
![imagen](https://github.com/frangoteam/FUXA/assets/6344456/c4081bd4-77d1-40e3-b407-c9992d556b9c)
Hi, there you can see a single history line, showing several phases of a tank clean process. This is done using a python script and a graphing library called bokeh. Each square color block is related to a value in the process variable. The process in the PLC is usually something like
 
Am I in stage 100 (process_var=100) and certain amount of flow is reached? then I write 200 in process_var
Am I in stage 200 (process_var=200) and a certain level is reached in the tank? then I write 300 in process_var
.....
.....
Once the sequence is coded, you can connect al motors, valves etc to a certain values of the precess_var, so that
every actuator will function only in the designed stage(s)

So if you assign every registered value of process var to a only one color, then when you plot it in a trend you get somthing
like a historic line. I saw at uPlot examples a close case:
https://leeoniya.github.io/uPlot/demos/timeline-discrete.html

The other case is how to represent digital values, because digital signals (proximity switches, etc etc ) are usually very important for all precesses, but for some extrange reason, the trending engines of scadas and HMIs doesn't have a way to reflect digital signals. I do that in weintek HMIs using analog traces, you only need to scale the trace to two possible values: Suppouse you have a trend scaled 0-100, then For logic 0 of a switch, the trace could be = 50 and for logic 1 the value sould be 55, then you get a squere wave horizontally centered in the trace. But it is because weintek software only let analog traces. whit a uPlot library i think one can plot direcly digital traces using bars or something like that.

**Reply by unocelli** [reply]
Hi, I'm not sure why the current settings are not enough for an analysis with digital and analogue signals.
![image](https://github.com/frangoteam/FUXA/assets/48684550/7e949fae-9901-46ae-b72a-af093b9e718a)

**Reply by mariorl** [reply]
Hi, how did you get the green square trace upper from the blue one? points to different Y scales?

**Comment by mariorl** [comment]
![imagen](https://github.com/frangoteam/FUXA/assets/6344456/8dafa253-f684-4776-b246-f2f6335f280d)
Please note. the initial aim of this topic was about representig machine states as colors in a gantt chart, rather than representing digital values. I mean I mentioned digital traces as a big help once you get the gantt chart. Having said that, I don't know how you get digital traces in the way sown in your previous post, looks very nice. I tested on my Fuxa server upon read your post, and didn't get this kind of plot. In order to explain the concept of digital signal analisys, attached you can see a capture o a very simple PLC program, and a weintek HMI tracing the 3 inputs and the single output. The digital traces are not overlapped in the X axis to improve clarity. The really important thing is that usually  you can't watch E2.0, E2.1 and E2.2 at same time and know what was the last input signal who turned the A2.0 output ON, Often the signals involved in a PLC logical operation are physically far ones from others. But with this kind of plots you can. Thanks and regards!
